// ttk-epay-nodejs-client/src/models.js
// models.js

/**
 * Class representing an Invoice
 */
class Invoice {
    /**
     * Create an Invoice instance
     * @param {Object} data - Invoice data
     * @param {number} [data.id] - Invoice ID
     * @param {string} [data.orderId] - Order ID
     * @param {string} [data.customerName] - Customer name
     * @param {string} [data.customerEmail] - Customer email
     * @param {number} [data.amount] - Invoice amount
     * @param {string} [data.currency] - Invoice currency
     * @param {string} [data.status] - Invoice status
     * @param {Date} [data.createdAt] - Creation date
     * @param {string} [data.description] - Invoice description
     * (Add other properties as needed based on your API requirements)
     */
    constructor(data = {}) {
      this.id = data.id || null;
      this.orderId = data.orderId || null;
      this.customerName = data.customerName || '';
      this.customerEmail = data.customerEmail || '';
      this.amount = data.amount || 0;
      this.currency = data.currency || 'DZD';
      this.status = data.status || 'PENDING';
      this.createdAt = data.createdAt || new Date();
      this.description = data.description || '';
      // Add other properties as needed
    }
  }
  
  /**
   * Class representing an InvoiceDto for payment processing
   */
  class InvoiceDto {
    /**
     * Create an InvoiceDto instance
     * @param {Object} data - Invoice data
     * @param {string} [data.orderId] - Order ID
     * @param {string} [data.customerName] - Customer name
     * @param {string} [data.customerEmail] - Customer email
     * @param {number} [data.amount] - Invoice amount
     * @param {string} [data.currency] - Invoice currency
     * @param {string} [data.successUrl] - Success redirect URL
     * @param {string} [data.failUrl] - Failure redirect URL
     * @param {string} [data.description] - Payment description
     * (Add other properties as needed based on your API requirements)
     */
    constructor(data = {}) {
      this.orderId = data.orderId || null;
      this.customerName = data.customerName || '';
      this.customerEmail = data.customerEmail || '';
      this.amount = data.amount || 0;
      this.currency = data.currency || 'DZD';
      this.successUrl = data.successUrl || '';
      this.failUrl = data.failUrl || '';
      this.description = data.description || '';
      // Add other properties as needed
    }
  }
  
  module.exports = {
    Invoice,
    InvoiceDto
  };
